File: leetcode/go/948.bag-of-tokens.go
func bagOfTokensScore(tokens []int, power int) int {
    slices.Sort(tokens)
    score := 0
    max_score := 0
    i := 0
    j := len(tokens)-1
    for i <= j {
        if power >= tokens[i] {
            power -= tokens[i];
            score++
            i++
        } else if score > 0 {
            power += tokens[j];
            score--
            j--
        } else {
            break;
        }
        max_score = max(max_score, score)
    }
    return max_score
}

File: leetcode/rust-leetcode/src/948.bag-of-tokens.rs
impl Solution {
    pub fn bag_of_tokens_score(mut tokens: Vec<i32>, mut power: i32) -> i32 {
        if tokens.len() == 0 {
            return 0;
        }
        let mut max_score = 0i32;
        
        tokens.sort();
        let mut i:usize = 0;
        let mut j = tokens.len() - 1;

        let mut score = 0i32;

        while i <= j {
            if power >= tokens[i] {
                power -= tokens[i];
                score += 1;
                i += 1;
            } else if score > 0 {
                power += tokens[j];
                score -= 1;
                j -= 1;
            } else {
                break;
            }
            max_score = max_score.max(score);
        }
        
        max_score
    }
}
